python/foxglove-sdk/src/websocket_server.rs

```diff
@@ -0,0 +1,254 @@
+/// :param host: The host to bind to.
+/// :param port: The port to bind to.
+/// :param capabilities: A list of capabilities to advertise to clients.
+/// :param server_listener: A Python object that implements the :py:class:`ServerListener` protocol.
+/// :param supported_encodings: A list of encodings to advertise to clients.
+///    Foxglove currently supports "json", "ros1", and "cdr" for client-side publishing.
+///
+/// To connect to this server: open Foxglove, choose "Open a new connection", and select Foxglove
+/// WebSocket. The default connection string matches the defaults used by the SDK.
+#[pyfunction]
+#[pyo3(signature = (*, name = None, host="127.0.0.1", port=8765, capabilities=None, server_listener=None, supported_encodings=None))]
+pub fn start_server(
+    py: Python<'_>,
+    name: Option<String>,
+    host: &str,
+    port: u16,
+    capabilities: Option<Vec<PyCapability>>,
+    server_listener: Option<Py<PyAny>>,
+    supported_encodings: Option<Vec<String>>,
+) -> PyResult<PyWebSocketServer> {
+    let session_id = time::SystemTime::now()
+        .duration_since(time::UNIX_EPOCH)
+        .expect("Failed to create session ID; invalid system time")
+        .as_millis()
+        .to_string();
+
+    let mut server = WebSocketServer::new()
+        .session_id(session_id)
+        .bind(host, port);
+
+    if let Some(py_obj) = server_listener {
+        let listener = PyServerListener::new(py_obj);
+        server = server.listener(Arc::new(listener));
+    }
+
+    if let Some(name) = name {
+        server = server.name(name);
+    }
+
+    if let Some(capabilities) = capabilities {
+        server = server.capabilities(capabilities.into_iter().map(PyCapability::into));
+    }
+
+    if let Some(supported_encodings) = supported_encodings {
+        server = server.supported_encodings(supported_encodings);
+    }
+
+    let handle = py
+        .allow_threads(|| server.start_blocking())
+        .map_err(PyFoxgloveError::from)?;
+
+    Ok(PyWebSocketServer(Some(handle)))
+}
+
+/// A live visualization server. Obtain an instance by calling :py:func:`start_server`.
+#[pyclass(name = "WebSocketServer", module = "foxglove")]
+pub struct PyWebSocketServer(pub Option<WebSocketServerBlockingHandle>);
+
+#[pymethods]
+impl PyWebSocketServer {
+    pub fn stop(&mut self, py: Python<'_>) {
+        if let Some(server) = self.0.take() {
+            py.allow_threads(|| server.stop())
+        }
+    }
+
+    /// Sets a new session ID and notifies all clients, causing them to reset their state.
+    /// If no session ID is provided, generates a new one based on the current timestamp.
+    #[pyo3(signature = (session_id=None))]
+    pub fn clear_session(&self, session_id: Option<String>) {
+        let Some(server) = &self.0 else {
+            return;
+        };
+        server.clear_session(session_id);
+    }
+
+    /// Publishes the current server timestamp to all clients.
+    pub fn broadcast_time(&self, timestamp_nanos: u64) {
```

Comment:
Is this from another PR that you used as a base? Not a big deal, I'll review that one next. You can figure out the logistics around merging them.

Reply:
Sorry, this is a wonky diff because I changed the conditionals to early returns; was added in #224.